iree/compiler/Conversion/LinalgToSPIRV/KernelDispatchUtils.cpp

```diff
@@ -153,5 +163,267 @@ LogicalResult createNumWorkgroupsFromLinearizedResultShape(
+    // shared memory available (maybe). For now, just hard-wire it.
+    tileSizeK = 32;
+  }
+  assert(tileSizes.empty());
+  SmallVector<int64_t, 4> ts = {nBatchesPerWorkitem,
+                                nRowsPerWorkitem * workgroupSize[1],
+                                nColsPerWorkitem * workgroupSize[0], tileSizeK};
+  tileSizes.emplace_back(std::move(ts));
+  return success();
+}
+
+/// The size of the co-operative matrix multiply operations on the device.
+// TODO(#3131): This needs to be queried from the device.
+Optional<std::array<int64_t, 3>> getCooperativeMatmulSubgroupSize(
+    Type dataType, Type accumulatorType) {
+  if (dataType.isa<IntegerType>() && dataType.isInteger(8) &&
+      accumulatorType.isInteger(32)) {
+    return std::array<int64_t, 3>{8, 8, 32};
+  }
+  if (dataType.isa<FloatType>() && dataType.isF16() &&
+      (accumulatorType.isF32() || accumulatorType.isF16())) {
+    return std::array<int64_t, 3>{8, 8, 16};
+  }
+  return {};
+}
+
+/// Launch configuration for using spv.CooperativeMatrixMulAddNV
+/// operations. Needs two levels of tiling.
+static LogicalResult getConfigForCooperativeMatmul(
+    linalg::MatmulOp op, spirv::ResourceLimitsAttr resourceLimits,
+    TileSizesListType &tileSizes, std::array<int64_t, 3> &workgroupSize,
+    std::array<int64_t, 3> &numSubgroups) {
+  auto targetEnv = spirv::TargetEnv(spirv::lookupTargetEnv(op));
+  if (!targetEnv.allows(spirv::Capability::CooperativeMatrixNV) ||
+      !targetEnv.allows(spirv::Extension::SPV_NV_cooperative_matrix))
+    return failure();
+
+  ShapedType lhsType = op.getOperand(0).getType().cast<ShapedType>();
+  ArrayRef<int64_t> lhsShape = lhsType.getShape();
+  ShapedType rhsType = op.getOperand(1).getType().cast<ShapedType>();
+  ArrayRef<int64_t> rhsShape = rhsType.getShape();
+  ShapedType outputType = op.getOperand(2).getType().cast<ShapedType>();
+
+  Optional<std::array<int64_t, 3>> coopMatmulSize =
+      getCooperativeMatmulSubgroupSize(lhsType.getElementType(),
+                                       outputType.getElementType());
+  if (!coopMatmulSize) return failure();
+
+  // Check that the matmul sizes are a multiple of the tilesize.
+  auto isMultipleOf = [](int64_t s, int64_t ts) {
+    return !ShapedType::isDynamic(s) && (s % ts) == 0;
+  };
+  if (!isMultipleOf(lhsShape[0], (*coopMatmulSize)[0]) ||
+      !isMultipleOf(rhsShape[1], (*coopMatmulSize)[1]) ||
+      !isMultipleOf(lhsShape[1], (*coopMatmulSize)[2]) ||
+      !isMultipleOf(rhsShape[0], (*coopMatmulSize)[2]))
+    return failure();
+
+  // TODO(ravishankarm, antiagainst): For now hardwire the subgroup size.
+  const int64_t subgroupSize = 32;
+  unsigned maxWorkgroupSize =
+      resourceLimits.max_compute_workgroup_invocations().getInt();
+  std::tie(numSubgroups[0], numSubgroups[1]) =
+      distributeProcs2D(maxWorkgroupSize / subgroupSize);
+  numSubgroups[2] = 1;
+  const int64_t numVecMatmulPerSubgroupX = 1;
+  const int64_t numVecMatmulPerSubgroupY = 1;
```

Comment:
I assume we'll want a higher number. Do you plan to handle it in a future PR? Maybe add a TODO?

Reply:
Added TODO.